File: phase_stepping_fit.py
#!/usr/bin/env python
# encoding: utf-8

"""Compare the least squares fit of synusoidal functions with the fourier
component analysis.

"""

from __future__ import division, print_function

import importlib
import numpy as np
from numpy import sin, cos, pi

from commandline_parser import commandline_parser

def cvphi_to_cab(c, v, phi):
    """convert the
    intensity, visibility, phase shift parameters to intensity and two
    angular parameters a, b such that:

    c[1 + v cos(px - phi)] = c + a cos(px) + b sin(px)

    the transformation gives:

    c = c
    a = vc cos(phi)
    b = vc sin(phi)

        >>> from math import pi
        >>> cvphi_to_cab(0, 0, 0)
        (0, 0.0, 0.0)
        >>> cvphi_to_cab(0.5, 1, 0)
        (0.5, 0.5, 0.0)
    """
    a = v * c * cos(phi)
    b = v * c * sin(phi)
    return c, a, b

def phase_stepping_curve(c, v, phi, n):
    """Return the phase stepping curve sampled over one period
    with average c, visibility v, shift phi and n steps.

        >>> phase_stepping_curve(0.5, 1, 0, 4)
        array([ 1. ,  0.5,  0. ,  0.5])
    """
    p = 2 * pi / n #period
    xs = np.arange(n)
    angles = p * xs + phi
    return c * (1 + v * np.cos(angles))

def least_squares_fit(phase_stepping_curve):
    """Fit y = c + a cos(px) + b sin(px) using
    the least squares analytical method.
    see Statistical Methods in Experimental Physics, p. 182, 8.4

    """
    n = phase_stepping_curve.shape[0]
    xs = np.arange(n)
    p = 2 * pi / n #period
    angles = p * xs
    A = np.vstack([np.ones_like(angles), np.cos(angles), -np.sin(angles)]).T
    c, a, b = np.linalg.lstsq(A, phase_stepping_curve)[0]
    return c, a, b

def fourier_fit(phase_stepping_curve):
    """Fit y = c + a cos(px) + b sin(px) using
    the fourier transform.

    """
    n = phase_stepping_curve.shape[0]
    transformed = np.fft.rfft(phase_stepping_curve)
    c = transformed[0].real / n
    a = transformed[1].real / (n / 2)
    b = transformed[1].imag / (n / 2)
    return c, a, b

def main(args):
    """Calculate the three parameters a, b, c in the phase stepping curve:
    y = c + a cos(px) + b sin(px) with the least squares and the fourier
    component analysis.

    Return the original parameters, the least squares estimates and the
    fourier estimates.
    """
    constant = args.constant
    phase = args.phase
    visibility = args.visibility
    steps = args.steps
    noise = args.noise
    noise_module = importlib.import_module("noise_types")
    noise_function = getattr(noise_module, noise)
    original_pars = cvphi_to_cab(constant, visibility, phase)
    curve = phase_stepping_curve(
            constant, visibility, phase, steps)
    noisy_curve = curve + noise_function(curve)
    least_squares_pars = least_squares_fit(noisy_curve)
    fourier_pars = fourier_fit(noisy_curve)
    return original_pars, least_squares_pars, fourier_pars

if __name__ == '__main__':
    args = commandline_parser.parse_args()
    original_pars, least_squares_pars, fourier_pars = main(args)
    print("original parameters:", original_pars)
    print("least squares fit:", least_squares_pars)
    print("fourier analysis:", fourier_pars)

File: phase_stepping_fit.R
#!/usr/bin/env Rscript

library(argparse)
library(gridExtra)
library(plyr)
library(ggplot2)

phase_stepping_curve = function(c, v, phi, n) {
    #Return the phase stepping curve sampled over one period
    #with average c, visibility v, shift phi and n steps.
    p = 2 * pi / n
    xs = 0:(n - 1)
    angles = p * xs + phi
    return(c * (1 + v * cos(angles)))
}

commandline_parser = ArgumentParser(
        description="test various fit methods for the phase stepping curves.") 

commandline_parser$add_argument('-n',
            type='integer', nargs='?', default=100,
            help='number of comparisons')
commandline_parser$add_argument('-v', '--visibility',
            type='double', nargs='?', default=0.8,
            help='visibility')
commandline_parser$add_argument('-p', '--phase',
            type='double', nargs='?', default=0,
            help='phase')
commandline_parser$add_argument('-c', '--constant',
            type='double', nargs='?', default=40000,
            help='average intensity')
commandline_parser$add_argument('-s', '--steps',
            type='integer', nargs='?', default=15,
            help='phase steps')

args = commandline_parser$parse_args()
tries = args$n
constants = rep(args$c, tries)
phases = rep(args$p, tries)
visibilities = rep(args$v, tries)
steps = rep(args$s, tries)

curves = Map(phase_stepping_curve,
             constants, visibilities, phases, steps)

noisy_curves = lapply(curves,
                      function(x) unlist(lapply(x, rpois, n=1)))

angles = Map(function(n) 2 * pi / n * (0:(n - 1)), steps)

fit_matrices = Map(cbind,
                   Map(cos, angles),
                   Map(sin, angles))

fits = Map(lsfit, fit_matrices, noisy_curves)

#fit with inverse of the square of the noise as weight
weights = lapply(noisy_curves, function(x) 1 / x)
weighted_fits = Map(lsfit, fit_matrices, noisy_curves, wt=weights)
weighted_coefficients = Map(as.numeric,
                            Map(get, "coefficients", weighted_fits))

coefficients = Map(as.numeric, Map(get, "coefficients", fits))
coefficients = data.frame(do.call(rbind, coefficients), row.names=NULL)
weighted_coefficients = data.frame(do.call(rbind, weighted_coefficients),
                            row.names=NULL)
original = data.frame(constant=constants, phase=phases, visibility=visibilities)
original$type = factor("original")
unweighted_fit = data.frame(constant=coefficients$X1,
                            phase=atan(-coefficients$X3 / coefficients$X2),
                            visibility=sqrt(
        coefficients$X2^2 + coefficients$X3^2) / coefficients$X1)
unweighted_fit$type = factor("unweighted")
weighted_fit = data.frame(constant=weighted_coefficients$X1,
            phase=atan(-weighted_coefficients$X3 / weighted_coefficients$X2),
            visibility=(sqrt(
                weighted_coefficients$X2^2 + weighted_coefficients$X3^2) /
                weighted_coefficients$X1))
weighted_fit$type = factor("weighted")
fits = rbind(unweighted_fit, weighted_fit)
#print(fits)
#plot histograms
hist_constant = ggplot(fits, aes(constant, fill=type)) + geom_histogram(
                    alpha=0.5,
                    position="identity",
                    binwidth=5)
hist_phase = ggplot(fits, aes(phase, fill=type)) + geom_histogram(
                    alpha=0.5,
                    position="identity",
                    binwidth=0.00015)
hist_visibility = ggplot(fits, aes(visibility, fill=type)) + geom_histogram(
                    alpha=0.5,
                    position="identity",
                    binwidth=0.0001)
x11(width=20)
histograms = arrangeGrob(hist_constant, hist_phase, hist_visibility, nrow=1)
print(histograms)
ggsave(sprintf("histograms%.2f.pdf", args$v), histograms)

#statistical tests
#variances
print(var.test(weighted_fit$visibility, unweighted_fit$visibility))
print(var.test(weighted_fit$phase, unweighted_fit$phase))
print(var.test(weighted_fit$constant, unweighted_fit$constant))
#means
print(t.test(weighted_fit$visibility, unweighted_fit$visibility))
print(t.test(weighted_fit$phase, unweighted_fit$phase))
print(t.test(weighted_fit$constant, unweighted_fit$constant, var.equal=TRUE))
message("Press Return To Continue")
invisible(readLines("stdin", n=1))
